src/array.rs

```diff
@@ -2,126 +2,143 @@ use super::{Error, Miniconf, MiniconfMetadata};
+            &self,
+            index: &mut [usize],
+            topic: &mut heapless::String<TS>,
+        ) -> Option<()> {
+            let original_length = topic.len();
 
-            if write!(topic, "{}", index[0]).is_err() {
+            if index.is_empty() {
                 // Note: During expected execution paths using `into_iter()`, the size of the
-                // topic buffer is checked in advance to make sure this condition doesn't occur.
+                // index stack is checked in advance to make sure this condition doesn't occur.
                 // However, it's possible to happen if the user manually calls `recurse_paths`.
-                unreachable!("Topic buffer too short");
+                unreachable!("Index stack too small");
             }
 
-            if self[index[0]]
-                .recurse_paths(&mut index[1..], topic)
-                .is_some()
-            {
-                return Some(());
+            while index[0] < self.len() {
+                // Add the array index to the topic name.
+                if topic.len() > 0 && topic.push('/').is_err() {
+                    // Note: During expected execution paths using `into_iter()`, the size of the
+                    // topic buffer is checked in advance to make sure this condition doesn't occur.
+                    // However, it's possible to happen if the user manually calls `recurse_paths`.
+                    unreachable!("Topic buffer too short");
+                }
+
+                if write!(topic, "{}", index[0]).is_err() {
+                    // Note: During expected execution paths using `into_iter()`, the size of the
+                    // topic buffer is checked in advance to make sure this condition doesn't occur.
+                    // However, it's possible to happen if the user manually calls `recurse_paths`.
+                    unreachable!("Topic buffer too short");
+                }
+
+                if self[index[0]]
+                    .recurse_paths(&mut index[1..], topic)
+                    .is_some()
+                {
+                    return Some(());
+                }
+
+                // Strip off the previously prepended index, since we completed that element and need
+                // to instead check the next one.
+                topic.truncate(original_length);
+
+                index[0] += 1;
+                index[1..].iter_mut().for_each(|x| *x = 0);
             }
 
-            // Strip off the previously prepended index, since we completed that element and need
-            // to instead check the next one.
-            topic.truncate(original_length);
-
-            index[0] += 1;
-            index[1..].iter_mut().for_each(|x| *x = 0);
+            None
         }
+    };
+}
+
+impl<T: Miniconf, const N: usize> Miniconf for [T; N] {
+    impl_array!();
+}
 
-        None
-    }
+impl<T: Miniconf, const N: usize> Miniconf for heapless::Vec<T, N> {
```

Comment:
Hmm. I had thought `Vec` should be like `String`, i.e. atomic. Otherwise there is no way to ever change the length of a `Vec`.

Reply:
Or better tag individual members with `#[miniconf(atomic)]` in the enclosing `#[derive(Miniconf)]` struct (this is roughly how other crates like `thiserror` control the `derive`).

Reply:
But all this is a bigger redesign. I'd propose landing this PR with `String` an `Vec` being terminal and then spawning an issue for the redesign.

Reply:
Honestly, `Vec` is a whole different beast because of how versatile it is. I would recommend we limit this to just `String` and figure out how to handle `Vec` later.

Reply:
It would be a one-liner to add support for `Vec` in the same way as for `String`.

Reply:
Yes, but I don't think that's how we would actually want it implemented. In my opinion, a Vec is much like an Option, where you have a variable-length list of configurable settings, not something that should be mutated directly via external means.

Reply:
IMO it's the opposite. `Vec` should be atomic and thus trivially support variable length data. `array` can be non-terminal.
(That is before the potential refactor along the lines described above.)

But OK. Let's have this PR only cover `String`.